File: src/open/shape.ts
import { OColumnName } from './columnCache';

import type { ColumnCacheReader, ColumnCacheWriter, ColumnValue } from './columnCache';
import type { OProperties, OValue, Primitive, Properties, ValueArray } from '../vectorTile.spec';

//? Shapes exist solely to deconstruct and rebuild objects.
//?
//? Shape limitations:
//? - all keys are strings.
//? - all values are either:
//? - - primitive types: strings, numbers (f32, f64, u64, i64), true, false, or null
//? - - sub types: an array of a shape or a nested object which is itself a shape
//? - - if the sub type is an array, ensure all elements are of the same type
//? The interfaces below help describe how shapes are built by the user.

/**
 * Primitive types that can be found in a shape
 */
export type PrimitiveShapes = 'string' | 'f32' | 'f64' | 'u64' | 'i64' | 'bool' | 'null';

/** The Shape Object But the values can only be primitives */
export interface ShapePrimitive {
  [key: string]: PrimitiveShapes;
}

/**
 * Arrays may contain either a primitive or an object whose values are primitives
 */
export type ShapePrimitiveType = PrimitiveShapes | ShapePrimitive;

/**
 * Shape types that can be found in a shapes object.
 * Either a primitive, an array containing any type, or a nested shape.
 * If the type is an array, all elements must be the same type
 */
export type ShapeType = PrimitiveShapes | [ShapePrimitiveType] | Shape;

/** The Shape Object */
export interface Shape {
  [key: string]: ShapeType;
}

/**
 * Create shapes
 *
 * Used by Layer's and Feature's M-Values
 * Must be an object of key values
 * all keys will be the same, values will be different
 * A layer's Shape defines what the properties look like for every Feature in that layer
 * so we only have to store the properties and M-Value shape **once** per layer.
 * @param cache - the cache where all data is stored in a column format
 * @param shape - the shape object to encode
 * @returns - The index of where the shape was stored in the cache
 */
export function encodeShape(cache: ColumnCacheWriter, shape: Shape): number {
  // this will store a "shape" of numbers on how to rebuild the object
  const shapeStore: number[] = [];
  // encode the shape data
  _encodeShape(shape, shapeStore, cache);
  // return the index of the shape index
  return cache.addColumnData(OColumnName.shapes, shapeStore);
}
/**
 * Encodes a shape object into a format suitable for storing in a cache.
 * @param shape - the shape object to encode. Recursively encodes nested objects.
 * @param shapeStore - list of key (string) indices. Also includes if the value is an object or not.
 * @param cache - the cache where all data is stored in a column format.
 */
function _encodeShape(shape: ShapeType, shapeStore: number[], cache: ColumnCacheWriter): void {
  if (Array.isArray(shape)) {
    shapeStore.push(0);
    _encodeShape(shape[0], shapeStore, cache);
  } else if (typeof shape === 'object') {
    const entries = Object.entries(shape);
    shapeStore.push(encodeAttribute(1, entries.length));
    for (const [key, value] of entries) {
      // store key
      shapeStore.push(cache.addColumnData(OColumnName.string, key));
      _encodeShape(value, shapeStore, cache);
    }
  } else {
    shapeStore.push(encodeAttribute(2, shapePrimitiveToColumnName(shape)));
  }
}

/**
 * @param shapeIndex - the index to the key indices and whether the value is an object or not
 * @param cache - the cache where all data is stored in a column format
 * @returns - The shape object
 */
export function decodeShape(shapeIndex: number, cache: ColumnCacheReader): Shape {
  const shapeStore: number[] = cache.getColumn(OColumnName.shapes, shapeIndex);
  // duplicate the array to avoid modifying the original
  return _decodeShape(cache, [...shapeStore]) as Shape;
}

/**
 * @param cache - the cache where all data is stored in a column format
 * @param shapeStore - the shape data encoded as an array
 * @returns - The shape object
 */
function _decodeShape(cache: ColumnCacheReader, shapeStore: number[]): ShapeType {
  const attribute = decodeAttribute(shapeStore.shift() ?? 0);

  if (attribute.type === 0) {
    // Array
    return [_decodeShape(cache, shapeStore) as ShapePrimitiveType];
  } else if (attribute.type === 1) {
    // Object
    const length = attribute.countOrCol;
    const obj: Shape = {};
    for (let i = 0; i < length; i++) {
      const keyIndex = shapeStore.shift() ?? 0;
      const key: string = cache.getColumn(OColumnName.string, keyIndex);
      obj[key] = _decodeShape(cache, shapeStore);
    }
    return obj;
  } else {
    // Primitive value
    return columnNameToShapePrimitive(attribute.countOrCol);
  }
}

/**
 * @param value - the value to encode
 * @param shape - the shape of the value
 * @param cache - the cache where all data is stored in a column format
 * @returns - The index of where the value was stored in the cache
 */
export function encodeValue(value: OProperties, shape: Shape, cache: ColumnCacheWriter): number {
  const valueStore: (number | ColumnValue)[] = [];
  _encodeValue(value, shape, valueStore, cache);
  return cache.addColumnData(OColumnName.shapes, valueStore);
}

/**
 * @param value - the value to encode
 * @param shape - the shape of the value
 * @param valueStore - list of key (string) indices. Also includes if the value is an object or not
 * @param cache - the cache where all data is stored in a column format
 */
function _encodeValue(
  value: OValue,
  shape: ShapeType,
  valueStore: (number | ColumnValue)[],
  cache: ColumnCacheWriter,
): void {
  // we follow the rules of the shape
  if (Array.isArray(shape)) {
    value = value as ValueArray;
    valueStore.push(value.length);
    for (const v of value) {
      _encodeValue(v, shape[0], valueStore, cache);
    }
  } else if (typeof shape === 'object') {
    const keys = Object.keys(shape);
    value = value as OProperties;
    for (const key of keys) {
      // key stored already by shape
      _encodeValue(value?.[key], shape[key], valueStore, cache);
    }
  } else {
    if (shape === 'string') {
      valueStore.push(cache.addColumnData(OColumnName.string, (value as string) ?? ''));
    } else if (shape === 'u64') {
      valueStore.push(cache.addNumber((value as number) ?? 0, OColumnName.unsigned));
    } else if (shape === 'i64') {
      valueStore.push(cache.addNumber((value as number) ?? 0, OColumnName.signed));
    } else if (shape === 'f32') {
      valueStore.push(cache.addNumber((value as number) ?? 0, OColumnName.float));
    } else if (shape === 'f64') {
      valueStore.push(cache.addNumber((value as number) ?? 0, OColumnName.double));
    } else if (shape === 'bool') {
      valueStore.push(cache.addNumber(value ? 1 : 0, OColumnName.unsigned));
    }
  }
}

/**
 * @param valueIndex - the index of the encoded value in the cache
 * @param shape - the shape of the value to decode
 * @param cache - the cache where all data is stored in a column format
 * @returns The decoded value
 */
export function decodeValue(
  valueIndex: number,
  shape: Shape,
  cache: ColumnCacheReader,
): OProperties {
  const valueStore: number[] = cache.getColumn(OColumnName.shapes, valueIndex);
  // duplicate the array to avoid modifying the original
  return _decodeValue([...valueStore], shape, cache) as OProperties;
}

/**
 * @param valueStore - the encoded value data as an array
 * @param shape - the shape of the value to decode
 * @param cache - the cache where all data is stored in a column format
 * @returns The decoded value
 */
function _decodeValue(valueStore: number[], shape: ShapeType, cache: ColumnCacheReader): OValue {
  if (Array.isArray(shape)) {
    const length = valueStore.shift() ?? 0;
    const arr: ValueArray = [];
    for (let i = 0; i < length; i++) {
      arr.push(
        _decodeValue(valueStore, shape[0], cache) as Primitive | { [key: string]: Primitive },
      );
    }
    return arr;
  } else if (typeof shape === 'object') {
    const obj: OProperties = {};
    for (const key in shape) {
      obj[key] = _decodeValue(valueStore, shape[key], cache);
    }
    return obj;
  } else {
    if (shape === 'null') return null;
    const columnValue = valueStore.shift() ?? 0;
    if (shape === 'string') {
      return cache.getColumn(OColumnName.string, columnValue);
    } else if (shape === 'bool') {
      return cache.getColumn(OColumnName.unsigned, columnValue) !== 0;
    } else if (shape === 'u64') {
      return cache.getColumn(OColumnName.unsigned, columnValue);
    } else if (shape === 'i64') {
      return cache.getColumn(OColumnName.signed, columnValue);
    } else if (shape === 'f32') {
      return cache.getColumn(OColumnName.float, columnValue);
    } else {
      // f64
      return cache.getColumn(OColumnName.double, columnValue);
    }
  }
}

/**
 * A shape pair for stronger compression and decoding
 */
export interface ShapePair {
  /** The type (0 - array, 1 - object, 2 - value) */
  type: 0 | 1 | 2;
  /** the length if object or array; or the column to read from */
  countOrCol: number;
}

/**
 * @param type - 0 is array, 1 is object, 2 is value
 * @param countOrColname - the length of the object or array; if value, its the column name
 * - can match columns for [0-4] (string, u64, i64, f32, f64),
 * - or matches a type: 5 represents bool and null, 6 represents array, 7 represents object
 * @returns - the encoded message
 */
function encodeAttribute(type: 0 | 1 | 2, countOrColname: number): number {
  return (countOrColname << 2) + type;
}

/**
 * @param num - the column and index encoded together
 * @returns - the decoded message
 */
function decodeAttribute(num: number): ShapePair {
  return { type: (num & 0b11) as 0 | 1 | 2, countOrCol: num >> 2 };
}

/**
 * @param type - the primitive shape to convert
 * @returns - the column name corresponding to the shape primitive
 */
function shapePrimitiveToColumnName(type: PrimitiveShapes): OColumnName {
  if (type === 'string') return OColumnName.string;
  else if (type === 'u64') return OColumnName.unsigned;
  else if (type === 'i64') return OColumnName.signed;
  else if (type === 'f32') return OColumnName.float;
  else if (type === 'f64') return OColumnName.double;
  else if (type === 'bool') return 5;
  else return 6;
}

/**
 * @param columnName - the column name to convert
 * @returns - the primitive shape corresponding to the column name
 */
function columnNameToShapePrimitive(columnName: OColumnName): PrimitiveShapes {
  if (columnName === OColumnName.string) return 'string';
  else if (columnName === OColumnName.unsigned) return 'u64';
  else if (columnName === OColumnName.signed) return 'i64';
  else if (columnName === OColumnName.float) return 'f32';
  else if (columnName === OColumnName.double) return 'f64';
  else if (columnName === 5) return 'bool';
  else return 'null';
}

//? The Following are utility functions when the user doesn't pre-define the Properties/M-Value
//? Shapes to store:

/**
 * @param data - the data to create the shape from
 * @returns - the shape type we want to create based upon the data
 */
export function createShapeFromData(data: OProperties[] | Properties[]): Shape {
  const shape: Shape = {};
  for (const d of data) {
    for (const k in d) shape[k] = getShapesValueType(d[k]);
  }
  return shape;
}

/**
 * Update/Mutate the shape from the data provided
 * @param shape - the shape
 * @param data - the data to update the shape
 */
export function updateShapeFromData(shape: Shape, data: OProperties | Properties): void {
  for (const k in data) shape[k] = getShapesValueType(data[k]);
}

/**
 * @param value - conform to the rules of OValue
 * @returns - the shape type
 */
function getShapesValueType(value: OValue): ShapeType {
  if (Array.isArray(value)) {
    // If it's a number type and the first number is a u64 but the second is an i64 or f64?
    // otherwise just return the first case in the array
    const types = value.map(getShapesValueType);
    return [validateTypes(types) as ShapePrimitiveType];
  } else if (typeof value === 'object' && value !== null) {
    return createShapeFromData([value]);
  } else {
    return getPrimitiveType(value);
  }
}

/**
 * @param value - the primtive value to get the type from
 * @returns - the primitive type from the value
 */
function getPrimitiveType(value: string | number | boolean | null): PrimitiveShapes {
  const type = typeof value;
  if (type === 'string') {
    return 'string';
  } else if (type === 'number') {
    if (Number.isInteger(value)) {
      return (value as number) < 0 ? 'i64' : 'u64';
    } else {
      return 'f64';
    }
  } else if (type === 'boolean') {
    return 'bool';
  } else {
    return 'null';
  }
}

/**
 * This is primarily to check if the type is a primitive.
 * If the primitive is a number, find the "depth", the most complex is f64, then i64, then u64.
 * Otherwise, if the primitives don't match, throw an error.
 * If the type is NOT a primitive, ensure that all types in the array match
 * @param types - either a primitive type, array, or object
 * @returns - a single type from the list to validate the correct type to be parsed from values later
 */
export function validateTypes(types: ShapeType[]): ShapeType {
  if (typeof types[0] === 'string') {
    // first ensure all primitive types are the same (excluding numbers)
    let baseType = types[0] as PrimitiveShapes;
    const basePrim = isNumber(baseType);
    for (const type of types as PrimitiveShapes[]) {
      if (type !== baseType) {
        if (isNumber(type) && basePrim) {
          baseType = getHighestOrderNumber(baseType, type);
        } else {
          throw new Error('All types must be the same');
        }
      }
    }
    return baseType;
  } else {
    const typeCheck = types.every(
      (t) => typeof t === typeof types[0] && Array.isArray(t) === Array.isArray(types[0]),
    );
    if (!typeCheck) throw new Error('All types must be the same');
    return types[0];
  }
}

/**
 * given a primitive type, check if one of them is a "number" type
 * @param type - any primitive type
 * @returns - true if the type is a number
 */
function isNumber(type: PrimitiveShapes): boolean {
  return type === 'i64' || type === 'u64' || type === 'f64';
}

/**
 * @param typeA - either i64, u64, or f64
 * @param typeB - either i64, u64, or f64
 * @returns - the "highest order number" e.g. f64 > i64 > u64
 */
function getHighestOrderNumber(typeA: PrimitiveShapes, typeB: PrimitiveShapes): PrimitiveShapes {
  if (typeA === 'f64' || typeB === 'f64') {
    return 'f64';
  } else if (typeA === 'i64' || typeB === 'i64') {
    return 'i64';
  } else {
    return 'u64';
  }
}

File: rust/open/shape.rs
use crate::open::{ColumnCacheReader, ColumnCacheWriter, ColumnValue};
use alloc::{vec, vec::Vec};
use s2json::{
    PrimitiveShape, PrimitiveShapeType, PrimitiveValue, Shape, ShapePrimitive, ShapeType, Value,
    ValuePrimitive, ValuePrimitiveType, ValueType,
};
use serde::{Deserialize, Serialize};

//? Shape

// Shapes exist solely to deconstruct and rebuild objects.
//
// Shape limitations:
// - all keys are strings.
// - all values are either:
// - - primitive types: strings, numbers (f32, f64, u64, i64), true, false, or null
// - - sub types: an array of a shape or a nested object which is itself a shape
// - - if the sub type is an array, ensure all elements are of the same type
// The interfaces below help describe how shapes are built by the user.

trait PrimitiveShapeToStore {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter);
    fn decode(shape_store: &mut Vec<usize>) -> Self;
}
impl PrimitiveShapeToStore for PrimitiveShape {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, _cache: &mut ColumnCacheWriter) {
        shape_store.push(ShapePair::encode(ShapeDefinition::Primitive, self.into()).into());
    }

    fn decode(shape_store: &mut Vec<usize>) -> Self {
        let shape_pair = ShapePair::decode(shape_store.remove(0));
        shape_pair.count_or_col.into()
    }
}

trait ShapePrimitiveTypeToStore {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter);
    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self;
}
impl ShapePrimitiveTypeToStore for PrimitiveShapeType {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter) {
        match self {
            Self::Primitive(prim) => prim.encode(shape_store, cache),
            Self::NestedPrimitive(nested) => {
                shape_store.push(ShapePair::encode(ShapeDefinition::Object, nested.len()).into());
                for (key, value) in nested.iter() {
                    shape_store.push(cache.add_string(key.clone()).into());
                    value.encode(shape_store, cache);
                }
            }
        }
    }

    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self {
        let shape_pair = ShapePair::decode(store.remove(0));
        match shape_pair.p_type {
            ShapeDefinition::Primitive => {
                Self::Primitive(PrimitiveShape::from(shape_pair.count_or_col))
            }
            ShapeDefinition::Object => {
                let mut nested = ShapePrimitive::new();
                for _ in 0..shape_pair.count_or_col {
                    nested.insert(cache.get_string(store.remove(0)), PrimitiveShape::decode(store));
                }
                Self::NestedPrimitive(nested)
            }
            #[tarpaulin::skip]
            _ => panic!("Unknown shape definition: {:?}", shape_pair),
        }
    }
}

trait ShapeTypeToStore {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter);
    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self;
}
impl ShapeTypeToStore for ShapeType {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter) {
        match self {
            Self::Primitive(prim) => prim.encode(shape_store, cache),
            Self::Array(array) => {
                shape_store.push(0.into());
                array.first().unwrap().encode(shape_store, cache);
            }
            Self::Nested(nested) => {
                nested.encode(shape_store, cache);
            }
        }
    }

    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self {
        let code = store.remove(0);
        let shape_pair = ShapePair::decode(code);
        match shape_pair.p_type {
            ShapeDefinition::Primitive => {
                Self::Primitive(PrimitiveShape::from(shape_pair.count_or_col))
            }
            ShapeDefinition::Array => Self::Array(vec![PrimitiveShapeType::decode(store, cache)]),
            ShapeDefinition::Object => {
                // reinsert code because shape will check it again
                store.insert(0, code);
                Self::Nested(Shape::decode(store, cache))
            }
        }
    }
}

/// Encode/Decode a Shape to the column cache
pub trait ShapeToStore {
    /// Encode the shape
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter);
    /// Decode the shape
    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self;
}
impl ShapeToStore for Shape {
    fn encode(&self, shape_store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter) {
        shape_store.push(ShapePair::encode(ShapeDefinition::Object, self.len()).into());
        for (key, value) in self.iter() {
            shape_store.push(cache.add_string(key.clone()).into());
            value.encode(shape_store, cache);
        }
    }
    fn decode(store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self {
        let mut map = Shape::new();
        let shape_pair = ShapePair::decode(store.remove(0));
        if shape_pair.p_type != ShapeDefinition::Object {
            panic!("expected object shape")
        }
        for _ in 0..shape_pair.count_or_col {
            let key = cache.get_string(store.remove(0));
            let shape = ShapeType::decode(store, cache);
            map.insert(key, shape);
        }
        map
    }
}

/// Create shapes
///
/// Used by Layer's and Feature's M-Values
/// Must be an object of key values
/// all keys will be the same, values will be different
/// A layer's Shape defines what the properties look like for every Feature in that layer
/// so we only have to store the properties and M-Value shape **once** per layer.
pub fn encode_shape(shape: &Shape, cache: &mut ColumnCacheWriter) -> usize {
    // this will store a "shape" of numbers on how to rebuild the object
    let mut shape_store = Vec::<ColumnValue>::new();
    // encode the shape data
    shape.encode(&mut shape_store, cache);
    // return the index of the shape index
    cache.add_shapes(shape_store)
}

/// Decode shapes from the column cache using an index to find the shape encoding
pub fn decode_shape(shape_index: usize, cache: &mut ColumnCacheReader) -> Shape {
    let mut shape_store = cache.get_shapes(shape_index);
    // duplicate the array to avoid modifying the original
    Shape::decode(&mut shape_store, cache)
}

/// A shape pair for stronger compression and decoding
#[derive(Debug, Clone, PartialEq)]
pub struct ShapePair {
    /// The type (0 - array, 1 - object, 2 - value)
    pub p_type: ShapeDefinition,
    /// the length if object or array; or the column to read from
    pub count_or_col: usize,
}
impl ShapePair {
    /// encode a shape pair
    pub fn encode(p_type: ShapeDefinition, count_or_col: usize) -> usize {
        (count_or_col << 2) + p_type as usize
    }

    /// decode a shape pair
    pub fn decode(num: usize) -> ShapePair {
        ShapePair { p_type: (num & 0b11).into(), count_or_col: num >> 2 }
    }
}
impl From<usize> for ShapePair {
    fn from(value: usize) -> Self {
        ShapePair::decode(value)
    }
}

/// The type of shape we are decoding
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
pub enum ShapeDefinition {
    /// an array
    Array = 0,
    /// an object
    Object = 1,
    /// a primitive
    Primitive = 2,
}
impl From<usize> for ShapeDefinition {
    fn from(value: usize) -> Self {
        match value {
            0 => ShapeDefinition::Array,
            1 => ShapeDefinition::Object,
            _ => ShapeDefinition::Primitive, // 2
        }
    }
}

trait PrimitiveValueToStore {
    fn encode(
        &self,
        shape: &PrimitiveShape,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    );
    fn decode(
        shape: &PrimitiveShape,
        store: &mut Vec<usize>,
        cache: &mut ColumnCacheReader,
    ) -> Self;
}
impl PrimitiveValueToStore for PrimitiveValue {
    fn encode(
        &self,
        shape: &PrimitiveShape,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    ) {
        match (self, shape) {
            // string
            (PrimitiveValue::String(s), PrimitiveShape::String) => {
                store.push(cache.add_string(s.clone()).into())
            }
            // u64
            (PrimitiveValue::U64(u), PrimitiveShape::U64) => store.push(cache.add_u64(*u).into()),
            // i64
            (PrimitiveValue::U64(u), PrimitiveShape::I64) => {
                store.push(cache.add_i64(*u as i64).into())
            }
            (PrimitiveValue::I64(i), PrimitiveShape::I64) => store.push(cache.add_i64(*i).into()),
            // f32
            (PrimitiveValue::U64(u), PrimitiveShape::F32) => {
                store.push(cache.add_f32(*u as f32).into())
            }
            (PrimitiveValue::I64(i), PrimitiveShape::F32) => {
                store.push(cache.add_f32(*i as f32).into())
            }
            (PrimitiveValue::F32(f), PrimitiveShape::F32) => store.push(cache.add_f32(*f).into()),
            // f64
            (PrimitiveValue::U64(u), PrimitiveShape::F64) => {
                store.push(cache.add_f64(*u as f64).into())
            }
            (PrimitiveValue::I64(i), PrimitiveShape::F64) => {
                store.push(cache.add_f64(*i as f64).into())
            }
            (PrimitiveValue::F32(f), PrimitiveShape::F64) => {
                store.push(cache.add_f64(*f as f64).into())
            }
            (PrimitiveValue::F64(f), PrimitiveShape::F64) => store.push(cache.add_f64(*f).into()),
            // bool
            (PrimitiveValue::Bool(b), PrimitiveShape::Bool) => {
                store.push(cache.add_u64(if *b { 1 } else { 0 }).into())
            }
            // null
            (PrimitiveValue::Null, PrimitiveShape::Null) => {}
            #[tarpaulin::skip]
            _ => panic!("shape mismatch"),
        }
    }

    fn decode(
        shape: &PrimitiveShape,
        store: &mut Vec<usize>,
        cache: &mut ColumnCacheReader,
    ) -> Self {
        let col_val = store.remove(0);
        match shape {
            PrimitiveShape::String => PrimitiveValue::String(cache.get_string(col_val)),
            PrimitiveShape::U64 => PrimitiveValue::U64(cache.get_unsigned(col_val)),
            PrimitiveShape::I64 => PrimitiveValue::I64(cache.get_signed(col_val)),
            PrimitiveShape::F32 => PrimitiveValue::F32(cache.get_float(col_val)),
            PrimitiveShape::F64 => PrimitiveValue::F64(cache.get_double(col_val)),
            PrimitiveShape::Bool => PrimitiveValue::Bool(cache.get_unsigned(col_val) == 1),
            PrimitiveShape::Null => {
                // put the column back because null does not need to be decoded
                store.insert(0, col_val);
                PrimitiveValue::Null
            }
        }
    }
}

trait PrimitiveShapeTypeToStore {
    fn encode(
        &self,
        shape: &PrimitiveShapeType,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    );
    fn decode(
        shape: &PrimitiveShapeType,
        store: &mut Vec<usize>,
        cache: &mut ColumnCacheReader,
    ) -> Self;
}
impl PrimitiveShapeTypeToStore for ValuePrimitiveType {
    fn encode(
        &self,
        shape: &PrimitiveShapeType,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    ) {
        match (shape, self) {
            (
                PrimitiveShapeType::Primitive(shape_prim),
                ValuePrimitiveType::Primitive(value_prim),
            ) => {
                value_prim.encode(shape_prim, store, cache);
            }
            (
                PrimitiveShapeType::NestedPrimitive(shape_nest),
                ValuePrimitiveType::NestedPrimitive(value_nest),
            ) => {
                for (key, prim_shape) in shape_nest.iter() {
                    let val = value_nest.get(key).unwrap();
                    val.encode(prim_shape, store, cache);
                }
            }
            #[tarpaulin::skip]
            _ => panic!("shape and value do not match"),
        }
    }
    fn decode(
        shape: &PrimitiveShapeType,
        store: &mut Vec<usize>,
        cache: &mut ColumnCacheReader,
    ) -> Self {
        match shape {
            PrimitiveShapeType::Primitive(shape_prim) => {
                ValuePrimitiveType::Primitive(PrimitiveValue::decode(shape_prim, store, cache))
            }
            PrimitiveShapeType::NestedPrimitive(shape_nest) => {
                let mut map = ValuePrimitive::new();
                for (key, shape) in shape_nest.iter() {
                    map.insert(key.clone(), PrimitiveValue::decode(shape, store, cache));
                }
                ValuePrimitiveType::NestedPrimitive(map)
            }
        }
    }
}

/// Encode/Decode a ValueType to the column cache with a Shape describing how to store
pub trait ValueTypeToStore {
    /// Encode the value type into the store
    fn encode(
        &self,
        shape: &ShapeType,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    );
    /// Decode the value type from the store
    fn decode(shape: &ShapeType, store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self;
}
impl ValueTypeToStore for ValueType {
    fn encode(
        &self,
        shape: &ShapeType,
        store: &mut Vec<ColumnValue>,
        cache: &mut ColumnCacheWriter,
    ) {
        match (self, shape) {
            (ValueType::Primitive(val), ShapeType::Primitive(shape)) => {
                val.encode(shape, store, cache);
            }
            (ValueType::Array(vals), ShapeType::Array(shape)) => {
                // encode length
                store.push(ColumnValue::Number(vals.len()));
                for val in vals {
                    val.encode(&shape[0], store, cache);
                }
            }
            (ValueType::Nested(val), ShapeType::Nested(shape)) => {
                val.encode(shape, store, cache);
            }
            #[tarpaulin::skip]
            _ => panic!("shape and value do not match"),
        }
    }

    fn decode(shape: &ShapeType, store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self {
        match shape {
            ShapeType::Primitive(shape_prim) => {
                ValueType::Primitive(PrimitiveValue::decode(shape_prim, store, cache))
            }
            ShapeType::Array(shape_arr) => {
                let mut val = Vec::<ValuePrimitiveType>::new();
                let len = store.remove(0);
                for _ in 0..len {
                    val.push(ValuePrimitiveType::decode(&shape_arr[0], store, cache));
                }
                ValueType::Array(val)
            }
            ShapeType::Nested(shape) => ValueType::Nested(Value::decode(shape, store, cache)),
        }
    }
}

/// Encode/Decode a Value to the column cache with a Shape describing how to store
pub trait ValueToStore {
    /// Encode the value into the store
    fn encode(&self, shape: &Shape, store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter);
    /// Decode the value from the store
    fn decode(shape: &Shape, store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self;
}
impl ValueToStore for Value {
    fn encode(&self, shape: &Shape, store: &mut Vec<ColumnValue>, cache: &mut ColumnCacheWriter) {
        for (key, shape_type) in shape.iter() {
            let val = self.get(key).unwrap_or(&ValueType::default_from_shape(shape_type)).clone();
            val.encode(shape_type, store, cache);
        }
    }
    fn decode(shape: &Shape, store: &mut Vec<usize>, cache: &mut ColumnCacheReader) -> Self {
        let mut value = Self::new();
        for (key, shape_type) in shape.iter() {
            let val = ValueType::decode(shape_type, store, cache);
            value.insert(key.clone(), val);
        }
        value
    }
}

/// Encode a value to the column cache
pub fn encode_value(value: &Value, shape: &Shape, cache: &mut ColumnCacheWriter) -> usize {
    let mut value_store: Vec<ColumnValue> = vec![];
    value.encode(shape, &mut value_store, cache);
    cache.add_shapes(value_store)
}

/// Decode a value from the column cache
pub fn decode_value(value_index: usize, shape: &Shape, cache: &mut ColumnCacheReader) -> Value {
    let value_store = cache.get_shapes(value_index);
    Value::decode(shape, &mut value_store.clone(), cache)
}
